File: ai/linear_model.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) {
  stop("Usage: Rscript linear_model.R <filename> <x_column> <y_column>")
}

filename <- args[1]
x_col <- args[2]
y_col <- args[3]

cat("Loading File:", filename, "\n")
cat("Using", x_col, "for x and", y_col, "for y\n")

data <- read.csv(filename)
x <- data[[x_col]]
y <- data[[y_col]]

formula <- as.formula(paste(y_col, "~", x_col))
model <- lm(formula, data = data)

slope <- coef(model)[2]
intercept <- coef(model)[1]
r <- cor(x, y)
pred <- predict(model)
mse <- mean((y - pred)^2)

cat(sprintf("Slope: %.2f\n", slope))
cat(sprintf("Intercept: %.2f\n", intercept))
cat(sprintf("Correlation coefficient (Pearson's r): %.2f\n", r))
cat(sprintf("Mean Squared Error (MSE): %.2f\n", mse))

library(ggplot2)

plot <- ggplot(data, aes(x = .data[[x_col]], y = .data[[y_col]])) +
  geom_point(color = "red") +
  geom_smooth(method = "lm", color = "blue", se = FALSE) +
  annotate(
    "text",
    x = min(x) + 0.1,
    y = max(y) * 0.92,
    label = paste0(
      "Equation: y = ", round(slope, 2), "x + ", round(intercept, 2),
      "\nCorrelation coefficient (r): ", round(r, 2)
    ),
    hjust = 0,
    size = 3.5
  ) +
  ggtitle(paste(y_col, "vs", x_col)) +
  xlab(x_col) +
  ylab(y_col)

ggsave("regression_plot_r.png", plot, width = 8, height = 6)
cat("Plot saved to regression_plot_r.png\n")

File: ai/linear_model.py
#!/usr/bin/env python
"""Command-line linear regression: fit, diagnostics, annotated plot, and PNG export."""

import sys

import matplotlib.pyplot as plt
import pandas as pd
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error


def main():
    if len(sys.argv) != 4:
        print("Usage: python linear_model.py <filename> <x_column> <y_column>")
        sys.exit(1)

    filename = sys.argv[1]
    x_column = sys.argv[2]
    y_column = sys.argv[3]

    print("Loading File: {}".format(filename))
    print("Using {} for x and {} for y".format(x_column, y_column))

    dataset = pd.read_csv(filename)
    x = dataset[x_column]
    y = dataset[y_column]

    slope, intercept, r_value, _p_value, _std_err = linregress(x, y)
    y_pred = slope * x + intercept
    mse = mean_squared_error(y, y_pred)

    print("Slope: {:.2f}".format(slope))
    print("Intercept: {:.2f}".format(intercept))
    print("Correlation coefficient (Pearson's r): {:.2f}".format(r_value))
    print("Mean Squared Error (MSE): {:.2f}".format(mse))

    plt.scatter(x, y, color="red", label="Observed data")
    plt.plot(x, y_pred, color="blue", label="Fitted line")
    plt.text(
        x.min() + 0.1,
        y.max() * 0.92,
        "Equation: y = {:.2f}x + {:.2f}\nCorrelation coefficient (r): {:.2f}".format(
            slope, intercept, r_value
        ),
        fontsize=10,
        bbox={"facecolor": "white", "alpha": 0.8, "edgecolor": "gray"},
    )
    plt.title("{} vs {}".format(y_column, x_column))
    plt.xlabel(x_column)
    plt.ylabel(y_column)
    plt.legend()
    plt.tight_layout()
    plt.savefig("regression_plot_python.png")
    print("Plot saved to regression_plot_python.png")


if __name__ == "__main__":
    main()
